### ISSUE TITLE
Replacing native C types (Windows & Mac vs Linux)

### ISSUE THREAD

**Issue by Wodann** [rust-lang/rust-bindgen#1663]
Bindgen nicely replaces occurrences of native C types with Rust types. For example:
```C
#include <stdint.h>

typedef struct
{
    const uint8_t b[16];
} MunGuid;
```

results in:

```rust
#[repr(C)]
#[derive(Clone, Copy)]
pub struct Guid {
    pub b: [u8; 16usize],
}
```

It works. Lovely! One problem, once we run CI, it turns out that Ubuntu also generates this additional line

```rust
pub type __uint8_t = ::std::os::raw::c_uchar;
```

Even though bindgen is able to deduce that `uint8_t` is supposed to be `u8`, it still adds the `__uint8_t` type to the generated bindings.

One possible solution that I came up with to resolved this is to blacklist the `__uint8_t` type, but then bindgen stops automatically generating the `#[derive(Clone, Copy)]` attributes.

Is there a way to prevent the generation of "replaced" types?

**Comment by emilio** [comment]
Other code in standard library (or user code, really) may actually be using `__uint8_t` type.

I would have expected this to work if you run this with `--whitelist-type MunGuid`, but it doesn't and it also outputs the `__uint8_t` type.

So I think the main issue is that the edge from `uint8_t` to `__uint8_t` is accounted for, even though bindgen manually avoids that edge during code generation.

So basically something like this line:

https://github.com/rust-lang/rust-bindgen/blob/79b2b1042069ca6673afc706fca2207d715f6894/src/ir/ty.rs#L1233

May be able to check its name and avoid tracing the inner type in that case. I don't know how that works perf-wise, would need some measuring.

**Comment by Ryan1729** [comment]
I got most of the way through the new issue template before noticing this seems to be the same issue I was experiencing. So I'm pasting that here. Hopefully the reduced testcase is useful to someone.

### Input C/C++ Header
`example.h`
```C
typedef unsigned int __uint32_t;
typedef __uint32_t uint32_t;

typedef struct Foo { uint32_t id; } Foo;
```

### Bindgen Invocation
```
$ bindgen example.h --blocklist-type '^__.*'
```

### Actual Results
```rust
/* automatically generated by rust-bindgen 0.59.2 */

#[repr(C)]
pub struct Foo {
    pub id: u32,
}
// Layout test elided
```

### Expected Results

I expected the struct to have `Clone`, and `Copy` derived. If I run `bindgen example.h` I get the following:
```rust
/* automatically generated by rust-bindgen 0.59.2 */

pub type __uint32_t = ::std::os::raw::c_uint;
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct Foo {
    pub id: u32,
}
```

These typedefs are based on what was in `stdint.h` on a Linux system
```C
typedef unsigned int __uint32_t;
typedef __uint32_t uint32_t;
```

This is related to https://github.com/rust-lang/rust-bindgen/issues/1215 but following one of the suggestions there to blocklist the undesired types causes `Copy` to not be derived, so I need to choose between having non-`Copy` types or an extra unneeded type.

**Comment by pvdrz** [comment]
#2287 made me realize about all this and I wonder if we could fix this from the rust side instead.

What I mean is that we could nive all this logic about special-casing `stdint.h` from `codegen` to `codegen::postprocessing` and replace all type aliases that use the `::std::os::raw::c_*` by their counterparts if the names are consistent or something.

This is guaranteed to just remove the aliases when they come from `stdint.h` or were user code done correctly so this:
```c
typedef unsigned char uint64_t;

uint64_t foo();
```
wont emit this by accident:
```rust
extern "C" {
    pub fn foo() -> u64;
}
```

**Comment by pvdrz** [comment]
An extra comment about this topic:

Replacing `uint32_t` by `u32` everywhere is very convenient but at the same time causes bindgen to emit invalid code from time to time (I'll concede that the example above is clearly just made to prove my point and is not realistic) and more importantly, forces us to sprinkle a bunch of special cases here and there. My strongest opinion about this is that correctness should precede convenience. 

Following that line of thought, we could add a `--replace-native-types` flag (name is a WIP :sweat_smile:) such that:
- when disabled, bindgen will emit the regular `std::os::raw::c_*` or `std::ffi::c_*` types.
- when enabled, bindgen will replace the `std::os::raw::c_*` or `std::ffi::c_*` types by its best guess.

I would keep this flag enabled by default and we could even generate extra tests checking for things like `assert_eq!(TypeId::of<c_int>(), TypeId::of::<i32>())` when the flag is enabled.
